packages/expo-file-system/ios/EXFileSystem/EXSessionTasks/EXSessionResumableDownloadTaskDelegate.m

```diff
@@ -0,0 +1,64 @@
+// Copyright 2015-present 650 Industries. All rights reserved.
+
+#import <EXFileSystem/EXSessionResumableDownloadTaskDelegate.h>
+
+@interface EXSessionResumableDownloadTaskDelegate ()
+
+@property (strong, nonatomic, readonly) EXDownloadDelegateOnWriteCallback onWrite;
+@property (strong, nonatomic, readonly) NSString *uuid;
+
+@end
+
+@implementation EXSessionResumableDownloadTaskDelegate
+
+- (instancetype)initWithSessionRegister:(id<EXSessionRegister>)sessionRegister
+                                resolve:(UMPromiseResolveBlock)resolve
+                                 reject:(UMPromiseRejectBlock)reject
+                           localFileUrl:(NSURL *)localFileUrl
+                              serverUrl:(NSURL *)serverUrl
+                              md5Option:(BOOL)md5Option
+                        onWriteCallback:(EXDownloadDelegateOnWriteCallback)onWrite
+                                   uuid:(NSString *)uuid;
+{
+  if (self = [self initWithSessionRegister:sessionRegister
+                                   resolve:resolve
+                                    reject:reject
+                              localFileUrl:localFileUrl
+                                 serverUrl:serverUrl
+                                 md5Option:md5Option]) {
+    _onWrite = onWrite;
+    _uuid = uuid;
+  }
+  
+  return self;
+}
+
+- (void)URLSession:(NSURLSession *)session downloadTask:(NSURLSessionDownloadTask *)downloadTask didWriteData:(int64_t)bytesWritten totalBytesWritten:(int64_t)totalBytesWritten totalBytesExpectedToWrite:(int64_t)totalBytesExpectedToWrite
+{
+  if (_onWrite) {
+    _onWrite(downloadTask, bytesWritten, totalBytesWritten, totalBytesExpectedToWrite);
+  }
+}
+
+- (void)URLSession:(NSURLSession *)session task:(NSURLSessionTask *)task didCompleteWithError:(NSError *)error
+{
+  if (error) {
+    // The task was paused by us. So, we shouldn't throw.
+    if (error.code == NSURLErrorCancelled) {
+      self.resolve([NSNull null]);
```

Comment:
Do we want to resolve with `null` even when the request is cancelled due to the app terminating?

Reply:
That means the user paused this task using js function.
How is it work?
- `FileSystem.createDownloadResumable()` creates an object which handles communication between the js side and native.
- next we can start task using `DownloadResumable.downloadAsync()` and this returns a promise
- `DownloadResumable.pauseAsync()` pauses this task and resolves two promises: itself and the one from `downloadAsync`

Reply:
Yes yes, I know of this scenario, I'm more wondering if it's the only scenario in which we get the `NSURLErrorCancelled` — what would happen if we `dealloc`ed `EXFileSystem`? IIRC `dealloc` would call `invalidateAndCancel` on both `NSURLSession`s — wouldn't it call the delegate chain and end up here, resolving to `null` (while in fact what we've done was dealloc the `EXFileSystem` instance, i.e. close/refresh the app)?

Reply:
I'll check that

Reply:
I've fixed it ;)